Sources/BSON/ExtendedJSONEncoder.swift

```diff
@@ -0,0 +1,39 @@
+import Foundation
+/// Facilitates the encoding of `Encodable` values into ExtendedJSON.
+public class ExtendedJSONEncoder {
+    /// The enum representing the two possible encoding modes.
+    public enum extendedJSONMode {
+        /// Canonical Extended JSON Format.
+        case canonical
+
+        /// Relaxed Extended JSON Format.
+        case relaxed
+    }
+
+    /// The options set on the encoder.
+    public struct Options {
+        /// Either Canonical or Relaxed Extended JSON Format to encode to.
+        public var mode: extendedJSONMode
+    }
+
+    /// Initialize an `ExtendedJSONEncoder`.
+    public init(options: ExtendedJSONEncoder.Options) {
```

Comment:
We talked about this on the design doc, but I am not quite sure where things ended up. looking at `BSONEncoder`, we don't actually support using a traditional options type directly. 

we do let you pass in a `CodingStrategyProvider` (under the label `options`, which is maybe kind of confusing in retrospect), but we actually store options directly on `BSONEncoder` as `var`s, and allow users to configure them (as well as `userInfo`, an option that is not part of `CodingStrategyProvider`) in that way as well.

I think I actually maybe lean toward a "no public options type" approach and just allowing configuration via mutating properties, but I don't feel that strongly. if we do keep a public struct, it seems like we should _also_ support directly mutating the properties directly on the encoder, for consistency w/ `BSONEncoder` and `JSONEncoder`.

we should probably also support the `userInfo` option, which we can pass through as the `userInfo` to the `BSONEncoder` (and maybe also `JSONEncoder`?) we use.

also, we should have notion of a default mode, which I think should be relaxed, that we use when a user does not specify one.

Reply:
@patrickfreed and I discussed this a bit and this is what he thought you landed on in the design doc discussion. I am fine with either approach

Reply:
Yeah, I don't feel too strongly about it either way, and given the weird situation with options on `BSONEncoder` perhaps its better to take the safer path and just include no options struct here.

>we should probably also support the userInfo option, which we can pass through as the userInfo to the BSONEncoder (and maybe also JSONEncoder?) we use.

This is a really good point, and something we definitely should do. I think we only need to pass it to the encoder that interfaces with `T`, which in this case is `BSONEncoder`. We also need to do this on the decoding end, so I filed SWIFT-959 for that work.

>also, we should have notion of a default mode, which I think should be relaxed, that we use when a user does not specify one.

Yep, and the spec recommends relaxed as the default, too. Though it also suggests having separate methods for canonical or relaxed, which we could consider as well.

Reply:
So, instead of passing anything into init, we can have 2 public variables: `mode`, and `userInfo`? 

> Yep, and the spec recommends relaxed as the default, too. Though it also suggests having separate methods for canonical or relaxed, which we could consider as well.

Does this mean having 2 encode methods? like `encodeRelaxed` and `encodeCanonical`? is that not confusing to people who expect an encoder to just have a method called `encode`?

Reply:
>So, instead of passing anything into init, we can have 2 public variables: mode, and userInfo?

Yep

>Does this mean having 2 encode methods? like encodeRelaxed and encodeCanonical? is that not confusing to people who expect an encoder to just have a method called encode?

Yeah, that's a good point. So it's probably best if we just stick to `encode` with a mode. We should probably document on `encode` the default value for `mode` and how users can configure it.

Reply:
> So, instead of passing anything into init, we can have 2 public variables: mode, and userInfo?

SGTM

> Does this mean having 2 encode methods? like `encodeRelaxed` and `encodeCanonical`? is that not confusing to people who expect an encoder to just have a method called `encode`?

I agree, I prefer the single method

Reply:
Okay I made those changes and added some details about the mode to the documentation.